libs/core/langchain_core/messages/utils.py

```diff
@@ -242,3 +248,604 @@ def convert_to_messages(
+                    id="blur",
+                ),
+            ]
+
+            merge_message_runs(messages)
+
+        .. code-block:: python
+
+            [
+                SystemMessage("you're a good assistant."),
+                HumanMessage("what's your favorite color\nwait your favorite food", id="foo",),
+                AIMessage(
+                    [
+                        "my favorite colo",
+                        {"type": "text", "text": "my favorite dish is lasagna"}
+                    ],
+                    tool_calls=[
+                        ToolCall({"name": "blah_tool", "args": {"x": 2}, "id": "123"),
+                        ToolCall({"name": "blah_tool", "args": {"x": -10}, "id": "456")
+                    ]
+                    id="baz"
+                ),
+            ]
+
+    """  # noqa: E501
+    if not messages:
+        return []
+    messages = convert_to_messages(messages)
+    merged: List[BaseMessage] = []
+    for msg in messages:
+        curr = msg.copy(deep=True)
+        last = merged.pop() if merged else None
+        if not last:
+            merged.append(curr)
+        elif isinstance(curr, ToolMessage) or not isinstance(curr, last.__class__):
+            merged.extend([last, curr])
+        else:
+            last_chunk = _msg_to_chunk(last)
+            curr_chunk = _msg_to_chunk(curr)
+            if isinstance(last_chunk.content, str) and isinstance(
+                curr_chunk.content, str
+            ):
+                last_chunk.content += "\n"
+            merged.append(_chunk_to_msg(last_chunk + curr_chunk))
+    return merged
+
+
+def trim_messages(
+    messages: Sequence[MessageLikeRepresentation],
+    *,
+    n_tokens: int,
+    token_counter: Union[
+        Callable[[List[BaseMessage]], int], Callable[[BaseMessage], int]
+    ],
+    strategy: Literal["first", "last"] = "last",
+    allow_partial: bool = False,
+    end_on: Optional[
+        Union[str, Type[BaseMessage], Sequence[Union[str, Type[BaseMessage]]]]
+    ] = None,
+    start_on: Optional[
+        Union[str, Type[BaseMessage], Sequence[Union[str, Type[BaseMessage]]]]
+    ] = None,
+    include_system: bool = False,
+    text_splitter: Optional[Callable[[str], List[str]]] = None,
+) -> List[BaseMessage]:
+    """Trim messages to be below a token count.
+
+    Args:
+        messages: Sequence of Message-like objects to trim.
+        n_tokens: Max token count of trimmed messages.
+        token_counter: Function for counting tokens of a BaseMessage or a list of
+            BaseMessage.
+        strategy: Strategy for trimming.
+            - "first": Keep the first <= n_count tokens of the messages.
+            - "last": Keep the last <= n_count tokens of the messages.
+        allow_partial: Whether to split a message if only part of the message can be
+            included.
+        end_on: The message type to end on. Can be specified as string names (e.g.
```

Comment:
Suggest explaining that the message list is trimmed to satisfy this (maybe that's obvious :slightly_smiling_face:)